File: overlay/map_markers.py
"""Map traffic-marker selection and hold-before-switch debouncing."""

from __future__ import annotations

from . import common as oc

MARKER_SLOTS = ("ahead", "behind", "leader")


def wrap_lap_delta(them: float, me: float) -> float:
    """Signed lap-distance delta (them - me), wrapped to (-0.5, 0.5]."""
    delta = them - me
    if delta > 0.5:
        delta -= 1.0
    elif delta < -0.5:
        delta += 1.0
    return delta


def marker_car_valid(
    idx: int,
    *,
    surface,
    on_pit_arr,
    pace_idxs: set[int],
    pit_surfaces: tuple[int, ...],
) -> bool:
    if idx in pace_idxs:
        return False
    if on_pit_arr is not None and idx < len(on_pit_arr) and on_pit_arr[idx]:
        return False
    if surface is None or idx >= len(surface):
        return False
    if surface[idx] in pit_surfaces:
        return False
    return surface[idx] == oc.TRK_ON_TRACK


def select_marker_candidates(
    player: int | None,
    lap_pct,
    surface,
    positions,
    *,
    pace_idxs: set[int],
    on_pit_arr,
    pit_surfaces: tuple[int, ...] = (oc.TRK_IN_PIT_STALL, oc.TRK_APPROACHING_PITS),
    alongside_zone: float = 0.004,
) -> dict[str, int | None]:
    """Raw CarIdx targets for ahead / behind / leader (no hold debounce).

    Ahead/behind are race-position neighbors (N-1 / N+1), not lap-distance.
    If that exact car is invalid (pits / off-track / pace), the slot is empty
    — there is no fallthrough to the next position.
    ``alongside_zone`` is accepted for call-site compatibility and ignored.
    """
    del alongside_zone  # race-position neighbors; kept for API compatibility
    out: dict[str, int | None] = {"ahead": None, "behind": None, "leader": None}
    if player is None or not lap_pct or player >= len(lap_pct):
        return out
    me = lap_pct[player]
    if me is None or me < 0:
        return out
    if not positions or player >= len(positions):
        return out
    my_pos = positions[player]
    try:
        my_pos = int(my_pos) if my_pos is not None else 0
    except (TypeError, ValueError):
        my_pos = 0
    if my_pos < 1:
        return out

    def _idx_at_pos(target: int) -> int | None:
        if target < 1:
            return None
        for idx, pos in enumerate(positions):
            try:
                p = int(pos) if pos is not None else 0
            except (TypeError, ValueError):
                continue
            if p != target or idx == player:
                continue
            # Found the car in that race position — valid or hide (no fallthrough).
            if marker_car_valid(
                idx, surface=surface, on_pit_arr=on_pit_arr,
                pace_idxs=pace_idxs, pit_surfaces=pit_surfaces,
            ):
                return idx
            return None
        return None

    out["ahead"] = _idx_at_pos(my_pos - 1)
    out["behind"] = _idx_at_pos(my_pos + 1)

    for idx, pos in enumerate(positions):
        try:
            p = int(pos) if pos is not None else 0
        except (TypeError, ValueError):
            continue
        if p != 1:
            continue
        if marker_car_valid(
            idx, surface=surface, on_pit_arr=on_pit_arr,
            pace_idxs=pace_idxs, pit_surfaces=pit_surfaces,
        ):
            out["leader"] = idx
        break
    return out


def apply_marker_hold(
    state: dict,
    candidate_idx: int | None,
    now: float,
    hold_sec: float,
    *,
    locked_valid: bool,
) -> int | None:
    """Update one slot's hold state; return the locked CarIdx to display."""
    if not locked_valid:
        state["locked"] = None

    if candidate_idx is None:
        state["pending"] = None
        state["pending_since"] = None
        return None

    locked = state.get("locked")
    if locked == candidate_idx:
        state["pending"] = None
        state["pending_since"] = None
        return locked

    pending = state.get("pending")
    if pending != candidate_idx:
        state["pending"] = candidate_idx
        state["pending_since"] = now

    pending_since = state.get("pending_since")
    if pending_since is not None and (now - pending_since) >= hold_sec:
        state["locked"] = candidate_idx
        state["pending"] = None
        state["pending_since"] = None
        return candidate_idx

    return locked if locked_valid else None


def resolve_traffic_markers(
    hold_states: dict[str, dict],
    candidates: dict[str, int | None],
    lap_pct,
    *,
    now: float,
    hold_sec: float,
    surface,
    on_pit_arr,
    pace_idxs: set[int],
    pit_surfaces: tuple[int, ...] = (oc.TRK_IN_PIT_STALL, oc.TRK_APPROACHING_PITS),
) -> dict[str, dict | None]:
    """Apply hold debounce and return idx + lap-% for each marker slot."""
    out: dict[str, dict | None] = {"ahead": None, "behind": None, "leader": None}
    for slot in MARKER_SLOTS:
        state = hold_states.setdefault(
            slot, {"locked": None, "pending": None, "pending_since": None})
        candidate = candidates.get(slot)
        locked = state.get("locked")
        locked_valid = (
            locked is not None
            and marker_car_valid(
                locked, surface=surface, on_pit_arr=on_pit_arr,
                pace_idxs=pace_idxs, pit_surfaces=pit_surfaces,
            )
        )
        idx = apply_marker_hold(
            state, candidate, now, hold_sec, locked_valid=locked_valid)
        if idx is not None and lap_pct and idx < len(lap_pct):
            pct = lap_pct[idx]
            if pct is not None and pct >= 0:
                out[slot] = {"idx": idx, "pct": float(pct)}
    return out


def fresh_hold_states() -> dict[str, dict]:
    return {slot: {"locked": None, "pending": None, "pending_since": None}
            for slot in MARKER_SLOTS}

File: overlay-rs/crates/gridglance-overlay/src/map_markers.rs
//! Map traffic-marker selection and hold-before-switch (Python `map_markers.py`).

use crate::telemetry::CarRow;
use std::collections::HashMap;

pub const MARKER_SLOTS: &[&str] = &["ahead", "behind", "leader"];

#[derive(Debug, Clone, Default)]
pub struct MarkerHoldSlot {
    pub locked: Option<i32>,
    pub pending: Option<i32>,
    pub pending_since: Option<f64>,
}

#[derive(Debug, Clone)]
pub struct TrafficMarker {
    pub idx: i32,
    pub pct: f32,
    pub label: String,
}

pub type HoldStates = HashMap<String, MarkerHoldSlot>;

pub fn fresh_hold_states() -> HoldStates {
    MARKER_SLOTS
        .iter()
        .map(|s| ((*s).to_string(), MarkerHoldSlot::default()))
        .collect()
}

pub fn marker_car_valid(car: &CarRow) -> bool {
    if car.is_pace_car {
        return false;
    }
    if car.on_pit || car.in_pit {
        return false;
    }
    car.on_track && car.lap_dist_pct >= 0.0
}

/// Raw CarIdx targets for ahead / behind / leader (no hold debounce).
/// Ahead/behind are race-position neighbors; no fallthrough if invalid.
pub fn select_marker_candidates(cars: &[CarRow]) -> HashMap<&'static str, Option<i32>> {
    let mut out: HashMap<&'static str, Option<i32>> = MARKER_SLOTS
        .iter()
        .map(|s| (*s, None))
        .collect();

    let Some(player) = cars.iter().find(|c| c.is_player) else {
        return out;
    };
    if player.lap_dist_pct < 0.0 || player.position < 1 {
        return out;
    }
    let my_pos = player.position;
    let player_idx = player.car_idx;

    let idx_at_pos = |target: i32| -> Option<i32> {
        if target < 1 {
            return None;
        }
        for c in cars {
            if c.position != target || c.car_idx == player_idx {
                continue;
            }
            // Found that race position — valid or hide (no fallthrough).
            return if marker_car_valid(c) {
                Some(c.car_idx)
            } else {
                None
            };
        }
        None
    };

    out.insert("ahead", idx_at_pos(my_pos - 1));
    out.insert("behind", idx_at_pos(my_pos + 1));

    for c in cars {
        if c.position != 1 {
            continue;
        }
        if marker_car_valid(c) {
            out.insert("leader", Some(c.car_idx));
        }
        break;
    }
    out
}

fn apply_marker_hold(
    state: &mut MarkerHoldSlot,
    candidate_idx: Option<i32>,
    now: f64,
    hold_sec: f64,
    locked_valid: bool,
) -> Option<i32> {
    if !locked_valid {
        state.locked = None;
    }

    let Some(candidate_idx) = candidate_idx else {
        state.pending = None;
        state.pending_since = None;
        return None;
    };

    if state.locked == Some(candidate_idx) {
        state.pending = None;
        state.pending_since = None;
        return state.locked;
    }

    if state.pending != Some(candidate_idx) {
        state.pending = Some(candidate_idx);
        state.pending_since = Some(now);
    }

    if let Some(since) = state.pending_since {
        if (now - since) >= hold_sec {
            state.locked = Some(candidate_idx);
            state.pending = None;
            state.pending_since = None;
            return Some(candidate_idx);
        }
    }

    if locked_valid {
        state.locked
    } else {
        None
    }
}

/// Apply hold debounce; return idx + lap-% + label for each marker slot.
pub fn resolve_traffic_markers(
    hold_states: &mut HoldStates,
    cars: &[CarRow],
    now: f64,
    hold_sec: f64,
) -> HashMap<&'static str, Option<TrafficMarker>> {
    let candidates = select_marker_candidates(cars);
    let mut out: HashMap<&'static str, Option<TrafficMarker>> = MARKER_SLOTS
        .iter()
        .map(|s| (*s, None))
        .collect();

    for &slot in MARKER_SLOTS {
        let state = hold_states.entry(slot.to_string()).or_default();
        let candidate = candidates.get(slot).copied().flatten();
        let locked = state.locked;
        let locked_valid = locked
            .and_then(|idx| cars.iter().find(|c| c.car_idx == idx))
            .is_some_and(marker_car_valid);
        let idx = apply_marker_hold(state, candidate, now, hold_sec, locked_valid);
        if let Some(idx) = idx {
            if let Some(c) = cars.iter().find(|c| c.car_idx == idx) {
                if c.lap_dist_pct >= 0.0 {
                    out.insert(
                        slot,
                        Some(TrafficMarker {
                            idx,
                            pct: c.lap_dist_pct,
                            label: c.car_number.clone(),
                        }),
                    );
                }
            }
        }
    }
    out
}

/// Slot name for a car idx (leader → ahead → behind; later overwrites).
pub fn marker_slots_by_idx(
    markers: &HashMap<&'static str, Option<TrafficMarker>>,
) -> HashMap<i32, &'static str> {
    let mut out = HashMap::new();
    for &slot in &["leader", "ahead", "behind"] {
        if let Some(Some(m)) = markers.get(slot) {
            out.insert(m.idx, slot);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn car(idx: i32, pos: i32, player: bool) -> CarRow {
        CarRow {
            car_idx: idx,
            position: pos,
            is_player: player,
            on_track: true,
            lap_dist_pct: 0.1 * idx as f32,
            car_number: format!("{idx}"),
            ..Default::default()
        }
    }

    #[test]
    fn selects_position_neighbors() {
        let cars = vec![
            car(0, 1, false),
            car(1, 2, false),
            car(2, 3, true),
            car(3, 4, false),
        ];
        let c = select_marker_candidates(&cars);
        assert_eq!(c["leader"], Some(0));
        assert_eq!(c["ahead"], Some(1));
        assert_eq!(c["behind"], Some(3));
    }

    #[test]
    fn no_fallthrough_when_ahead_in_pit() {
        let mut cars = vec![
            car(0, 1, false),
            car(1, 2, false),
            car(2, 3, true),
        ];
        cars[1].on_pit = true;
        cars[1].on_track = false;
        let c = select_marker_candidates(&cars);
        assert_eq!(c["ahead"], None);
        assert_eq!(c["leader"], Some(0));
    }
}
